lib/graphql/schema/warden.rb

```diff
@@ -126,8 +126,53 @@ def root_type_for_operation(op_name)
 
       private
 
+      def union_memberships(obj_type)
+        @unions ||=  read_through { |obj_type| @schema.union_memberships(obj_type) }
+        @unions[obj_type]
+      end
+
       def visible_field?(field_defn)
-        visible?(field_defn) && visible?(field_defn.type.unwrap)
+        visible?(field_defn) && visible_type?(field_defn.type.unwrap)
+      end
+
+      def visible_type?(type_defn)
+        return false unless visible?(type_defn)
+
+        if type_defn.kind.object? && has_abstract_type?(type_defn)
+          visible_implementation?(type_defn)
+        else
+          referenced_by_visible_members?(type_defn)
+        end
+      end
+
+      def visible_implementation?(type_defn)
+        if orphan?(type_defn)
+          member_or_implements?(type_defn)
+        else
+          referenced_by_visible_members?(type_defn) || member_or_implements?(type_defn)
+        end
+      end
+
+      def referenced_by_visible_members?(type_defn)
+        members = @schema.get_members_of_type(type_defn.unwrap.name)
+        # Types should be visible if there are
+        # no member definitions with that type
+        return true unless members.any?
+        # Only hide the type when there are fields
+        # but none are visible.
+        members.any? { |member| visible?(member) }
```

Comment:
Just a matter of taste but I thought I'd share, another approach is: 

```ruby 
# Show this type if: 
# - it has no references to begin with 
# - it has references, and some of them are still visible 
members.empty? || members.any? { |m| visible?(m) } 
```

Personally, I prefer it because: 
- The method has only one expression which returns, at it's at the end of the method body (no explicit `return` part-way through)
- There's no inverted conditional logic (no `unless`)

Of course, it's 👌  as is, just thought I'd share!